qiskit/transpiler/passes/basis/basis_translator.py

```diff
@@ -295,6 +248,110 @@ def replace_node(node, instr_map):
+        self, dag, qarg_indices, source_basis=None, qargs_local_source_basis=None
+    ):
+        if source_basis is None:
+            source_basis = set()
+        if qargs_local_source_basis is None:
+            qargs_local_source_basis = defaultdict(set)
+        for node in dag.op_nodes():
+            qargs = tuple(qarg_indices[bit] for bit in node.qargs)
+            if dag.has_calibration_for(node):
+                continue
+            # Treat the instruction as on an incomplete basis if the qargs are in the
+            # qargs_with_non_global_operation dictionary or if any of the qubits in qargs
+            # are a superset for a non-local operation. For example, if the qargs
+            # are (0, 1) and that's a global (ie no non-local operations on (0, 1)
+            # operation but there is a non-local operation on (1,) we need to
+            # do an extra non-local search for this op to ensure we include any
+            # single qubit operation for (1,) as valid. This pattern also holds
+            # true for > 2q ops too (so for 4q operations we need to check for 3q, 2q,
+            # and 1q operations in the same manner)
+            if qargs in self._qargs_with_non_global_operation or any(
+                frozenset(qargs).issuperset(incomplete_qargs)
+                for incomplete_qargs in self._qargs_with_non_global_operation
+            ):
+                qargs_local_source_basis[frozenset(qargs)].add((node.name, node.op.num_qubits))
+            else:
+                source_basis.add((node.name, node.op.num_qubits))
+            if isinstance(node.op, ControlFlowOp):
+                for block in node.op.blocks:
+                    block_dag = circuit_to_dag(block)
+                    source_basis, qargs_local_source_basis = self._extract_basis_target(
+                        block_dag,
+                        qarg_indices,
+                        source_basis=source_basis,
+                        qargs_local_source_basis=qargs_local_source_basis,
+                    )
+        return source_basis, qargs_local_source_basis
+
+
+# this could be singledispatchmethod and included in above class when minimum
+# supported python version=3.8.
+@singledispatch
+def _extract_basis(circuit):
+    return circuit
+
+
+@_extract_basis.register
+def _(dag: DAGCircuit):
+    for node in dag.op_nodes():
+        if not dag.has_calibration_for(node):
+            yield (node.name, node.op.num_qubits)
+        if isinstance(node.op, ControlFlowOp):
+            for block in node.op.blocks:
+                yield from _extract_basis(block)
+
+
+@_extract_basis.register
+def _(circ: QuantumCircuit):
+    for instr_context in circ.data:
+        instr, _, _ = instr_context
+        if not circ.has_calibration_for(instr_context):
+            yield (instr.name, instr.num_qubits)
+        if isinstance(instr, ControlFlowOp):
+            for block in instr.blocks:
+                yield from _extract_basis(block)
```

Comment:
This is a nice way of going about the type switching - nice and easy to read!

Reply:
Maybe we could also consider using multiple dispatch in general although currently it is outside the standard library.